e2e/qgb/qgb_network.go

```diff
@@ -0,0 +1,392 @@
+func (network QGBNetwork) StartBase() error {
+	err := network.Instance.
+		WithCommand([]string{"build", "core0"}).
+		Invoke().Error
+	if err != nil {
+		return err
+	}
+	err = network.Instance.
+		WithCommand([]string{"up", "-d", "--no-build", "core0"}).
+		Invoke().Error
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (network QGBNetwork) WaitForNodeToStart(rpcAddr string) error {
+	timeoutChan := time.After(5 * time.Minute)
+	for {
+		select {
+		case <-timeoutChan:
+			return fmt.Errorf("node %s not initialized in time", rpcAddr)
+		default:
+			trpc, err := http.New(rpcAddr, "/websocket")
+			if err != nil || trpc.Start() != nil {
+				fmt.Println("waiting for node to start...")
+				time.Sleep(5 * time.Second)
+				continue
+			}
+			return nil
+		}
+	}
+}
+
+func (network QGBNetwork) WaitForBlock(ctx context.Context, height int64) error {
+	err := network.WaitForNodeToStart(network.TendermintRPC)
+	if err != nil {
+		return err
+	}
+	trpc, err := http.New(network.TendermintRPC, "/websocket")
+	if err != nil {
+		return err
+	}
+	err = trpc.Start()
+	if err != nil {
+		return err
+	}
+	timeoutChan := time.After(5 * time.Minute)
+	for {
+		select {
+		case <-timeoutChan:
+			return fmt.Errorf("chain didn't reach height in time")
+		default:
+			status, err := trpc.Status(ctx)
+			if err != nil || status.SyncInfo.LatestBlockHeight < height {
+				fmt.Printf("current height: %d\n", status.SyncInfo.LatestBlockHeight)
+				time.Sleep(5 * time.Second)
+				continue
+			}
+			return nil
+		}
+	}
+}
+
+// WaitForOrchestratorToStart waits for the orchestrator having the celes address `accountAddress`
+// to sign the first data commitment (could be upgraded to get any signature, either valset or data commitment,
+// and for any nonce, but would require adding a new method to the querier. Don't think it is worth it now as
+// the number of valsets that will be signed is trivial and reaching 0 would be in no time).
+func (network QGBNetwork) WaitForOrchestratorToStart(ctx context.Context, accountAddress string) error {
+	querier, err := orchestrator.NewQuerier(network.CelestiaGRPC, network.TendermintRPC, nil)
+	if err != nil {
+		return err
+	}
+	defer querier.Stop()
+	timeoutChan := time.After(5 * time.Minute)
+	for {
+		select {
+		case <-timeoutChan:
+			return fmt.Errorf("orchestrator didn't start correctly")
+		default:
+			confirm, err := querier.QueryDataCommitmentConfirm(ctx, types.DataCommitmentWindow, 0, accountAddress)
+			if err == nil && confirm != nil {
+				return nil
+			}
+			fmt.Println("waiting for orchestrator to start ...")
+			time.Sleep(5 * time.Second)
+		}
+	}
+}
+
+func (network QGBNetwork) GetLatestDeployedQGBContract(ctx context.Context) (*wrapper.QuantumGravityBridge, error) {
+	return network.GetLatestDeployedQGBContractWithCustomTimeout(ctx, 5*time.Minute)
+}
+
+func (network QGBNetwork) GetLatestDeployedQGBContractWithCustomTimeout(
+	ctx context.Context,
+	timeout time.Duration,
```

Comment:
I definitely don't think we should change this in this PR, as its not even a nit pick just a different option, but we could simply rely on this context to cancel instead of a timeout channel. :shrug:

Reply:
I might have already commented about this, if so pls ignore

Reply:
Yes, we can actually rely on both, the context and the timeout, I am planning to implement it later as the current way this is done gives some issues when ctrl+c a running test... Will create an issue